File: day5/day5.py
#!/usr/bin/env python3

import sys

def react(data):
    i, j = 0, 1
    length = len(data)
    while i < length and j < length:
        if data[i] != data[j] and data[i].lower() == data[j].lower():
            data[i] = data[j] = "*"
            j += 1
            while i > 0 and data[i] == '*':
                i -= 1
        else:
            i = j
            j = i + 1

    data = [x for x in data if x != "*"]
    return data, len(data)

def part1(txt):
    _, length = react(list(txt))
    return length

def part2(txt):
    data, minlen = react(list(txt))
    letters = set(x.lower() for x in data)
    for letter in letters:
        _, l = react([x for x in data if x.lower() != letter])
        if minlen > l:
            minlen = l
    return minlen

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", part1(txt))
    print("Part2:", part2(txt))

File: day5/day5.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static size_t react(char *str)
{
	size_t removed = 0;
	size_t i;
	for (i = 0; str[i] == '*'; i++)
	{
		removed++;
	}
	size_t j;
	for (j = i + 1; str[j] == '*'; j++)
	{
		removed++;
	}
	size_t len = strlen(str);
	while (i < len && j < len)
	{
		if (str[i] != str[j] && tolower(str[i]) == tolower(str[j]))
		{
			removed++;
			str[i] = str[j] = '*';
			for (; str[j] == '*'; j++)
			{
				removed++;
			}
			while (i > 0 && str[i] == '*')
			{
				i--;
			}
		}
		else
		{
			for (i = j; str[i] == '*'; i++)
			{
				removed++;
			}
			for (j = i + 1; str[j] == '*'; j++)
			{
				removed++;
			}
		}
	}
	return len-removed;
}

static size_t part1(const char *str)
{
	char *dup = strdup(str);
	size_t rv = react(dup);
	free(dup);
	return rv;
}

static size_t part2(const char *str)
{
	char *copy1 = strdup(str);
	char *copy2 = strdup(str);

	/* first reduction */
	size_t minlen = react(copy1);
	for (int i = 'a'; i <= 'z'; i++)
	{
		for (char *src = copy1, *dst = copy2; *src; src++, dst++)
		{
			*dst = (tolower(*src) == i) ? '*' : *src;
		}
		size_t len = react(copy2);
		if (minlen > len)
		{
			minlen = len;
		}
	}
	free(copy2);
	free(copy1);
	return minlen;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s filename\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "File %s not found\n", argv[1]);
		return 1;
	}

	char *line = NULL;
	size_t bufsize = 0;
	if (getline(&line, &bufsize, input) == -1)
	{
		free(line);
		fclose(input);
		fprintf(stderr, "Cannot parse the data\n");
		return 1;
	}
	fclose(input);

	/* chop off the terminal \n */
	size_t linesize = strlen(line) - 1;
	line[linesize] = 0;

	printf("Part1: %zu\n", part1(line));
	printf("Part2: %zu\n", part2(line));
	free(line);
	return 0;
}
